Validate options in async_stream* functions

sabiwara opened the pull request:
This is a proposal for discussion.

Working with `async_stream*` functions can be a bit error prone and requires to constantly double check the doc (which often needs to jump to a separate function), since any typo in an option would fail either silently or much later down the lane when a timeout/exception happens.

The proposal is to validate these upfront instead of lazily, and with helpful error messages.

I also plan to add a dedicated type `t:Task.async_stream_options/1` following [this pattern](https://hexdocs.pm/elixir/1.16.0/GenServer.html#t:option/0), which could help check the spec in the IDE more easily. Thoughts?

sabiwara commented on lib/elixir/lib/task/supervised.ex:
```diff
@@ -189,20 +189,46 @@ defmodule Task.Supervised do
-  def stream(enumerable, acc, reducer, callers, mfa, options, spawn) do
-    next = &Enumerable.reduce(enumerable, &1, fn x, acc -> {:suspend, [x | acc]} end)
-    max_concurrency = Keyword.get(options, :max_concurrency, System.schedulers_online())
+  def validate_stream_options!(options) do
```
Note: this function is called immediately when building the stream, not when evaluating it.

sabiwara commented on lib/elixir/lib/task/supervised.ex:
```diff
@@ -189,20 +189,46 @@ defmodule Task.Supervised do
+        max_concurrency: System.schedulers_online(),
+        on_timeout: :exit,
+        timeout: 5000,
+        ordered: true,
+        zip_input_on_exit: false,
```
Note: I didn't enforce these are booleans since truthiness works, not sure if we should be more strict at the risk of breaking code.

josevalim commented:
I have approved the pull request code-wise but I am not sure if we should merge it.

To me, the main question is: we typically don't validate options in Elixir. If we start validating them, are we open to validating all of them? For example, are we going to validate the arguments to `String.split`?

I can see two issues:

1. performance implications in the work done to reject invalid options
2. potential breaking changes for users passing options through

Ideally we would have the type system do the work of validating these options, which will avoid both problems above. At the same time, maybe async_stream is complex (and expensive) enough to justify being an exception and adding validation to it today. Thoughts?

sabiwara commented:
> At the same time, maybe async_stream is complex (and expensive) enough to justify being an exception and adding validation to it today. Thoughts?

I had precisely the same concerns, but these were my thoughts for this case: complex and expensive enough to justify.

Another reason I went for it was that we already had some validation for [`max_concurrency`](https://github.com/elixir-lang/elixir/pull/13289/files#diff-7e6cea74edeabe7542bc9de72c3d9d314bfabae91d7cbe8e3dbae4ca81064c39R207-R209), although it was lazy.

> performance implications in the work done to reject invalid options

 I haven't benched it but my hunch is that if such a small overhead is of any significance, we probably shouldn't be using `async_stream` in the first place for this workload.
 
> potential breaking changes for users passing options through

We could start with `Keyword.validate/2` (without the bang) and a warning to avoid making this change breaking but let people find out?

>  If we start validating them, are we open to validating all of them? For example, are we going to validate the arguments to String.split?

I'd want to say no, this doesn't have to be a rule, I think this case justifies it (esp. because tends to fail later and in hard to test ways).

> Ideally we would have the type system do the work of validating these options

The type system can indeed be the answer for a more long term fix, which would also cover all other cases like `String.split` where runtime checks would be too expensive. But since it might not be right away, I figured I'd still suggest the runtime check in the meantime.

sabiwara commented:
OK I did some profiling with `fprof` on a contrived example and it seems it can have a measurable impact.


fprof -ing a contrieved example which does no work at all:

```elixir
def profilee do
  [1, 2, 3]
  |> Task.async_stream(& &1, ordered: false, timeout: 10)
  |> Enum.map(fn {:ok, res} -> res end)
end
```

```
                                                                   CNT    ACC (ms)    OWN (ms)
Total                                                              273       0.519       0.534
...
Keyword.validate!/2                                                  1       0.050       0.003
Keyword.validate/2                                                   1       0.047       0.002
Keyword.validate/5                                                   3       0.045       0.029
...
```

Even if this is probably unlikely to be representative of real world uses of async_stream, it means there's room for improvement.

sabiwara commented:
Pushed [e4ed509](https://github.com/elixir-lang/elixir/pull/13289/commits/e4ed509eb92ff4e85e885bbaf737388e02268c75) which just uses a map instead of `Keyword.validate!`.

We can add a [simple check for unknown keys](https://github.com/sabiwara/elixir_benches/commit/0cb1a9bc92019b569f7dfd1d891deef53479c47a#diff-867efc7b596b7b214a86e6c74c8527a2a860aa86561418dd082688c5d8874375R69-R75) as well without too much overhead (without checking dupes), WDYT?

josevalim commented:
The issue is that it can be a breaking change, so let’s focus on non-breaking changes that improve the UX!

sabiwara commented:
> The issue is that it can be a breaking change, so let’s focus on non-breaking changes that improve the UX!

The suggestion simply adds a warning so technically non breaking I think?

whatyouhide commented on lib/elixir/lib/task.ex:
```diff
@@ -293,6 +293,19 @@ defmodule Task do
+  functions from the `Task` and `Task.Supervisor` modules.
+  """
+  @typedoc since: "1.17.0"
+  @type async_stream_options :: [
```
Usually we tend to go with a type for the single option, like `async_stream_option() :: {:ordered, boolean} | ...` and then do `[async_stream_option()]` in the spec. See `GenServer.option/0` for example

sabiwara replied:
Yes, sorry I pushed too fast.
I also realized that there is a difference between `Task` (no `:shutdown`) and `Task.Supervisor` (has `:shutdown`), will clean this up.

sabiwara commented:
Fixed the `:shutdown` issue [d236b66](https://github.com/elixir-lang/elixir/pull/13289/commits/d236b661796d9542908a10dd5b4734a1ea189c8f) and moved the specs to https://github.com/elixir-lang/elixir/pull/13306 for easier review
